src/storage/sql/mod.rs

```diff
@@ -1 +1,319 @@
+                snippets::updated_at.eq(diesel::dsl::now),
+            ))
+            .returning(snippets::id)
+            .get_result::<i32>(conn);
+        let snippet_id = match result {
+            Ok(snippet_id) => snippet_id,
+            Err(DatabaseError(DatabaseErrorKind::UniqueViolation, _)) => {
+                return Err(StorageError::Duplicate {
+                    id: snippet.id.to_owned(),
+                })
+            }
+            Err(e) => return Err(StorageError::from(e)),
+        };
+
+        Ok(snippet_id)
+    }
+
+    fn upsert_changesets(
+        &self,
+        conn: &PgConnection,
+        snippet_id: i32,
+        snippet: &Snippet,
+    ) -> Result<(), StorageError> {
+        diesel::insert_into(changesets::table)
+            .values(
+                snippet
+                    .changesets
+                    .iter()
+                    .map(|c| {
+                        (
+                            changesets::snippet_id.eq(snippet_id),
+                            changesets::version.eq(c.version as i32),
+                            changesets::content.eq(&c.content),
+                            changesets::created_at.eq(diesel::dsl::now),
+                        )
+                    })
+                    .collect::<Vec<_>>(),
+            )
+            .on_conflict((changesets::snippet_id, changesets::version))
+            .do_update()
+            .set((
+                changesets::content.eq(upsert::excluded(changesets::content)),
+                changesets::updated_at.eq(diesel::dsl::now),
+            ))
+            .execute(conn)?;
+
+        Ok(())
+    }
+
+    fn upsert_tags(
+        &self,
+        conn: &PgConnection,
+        snippet_id: i32,
+        snippet: &Snippet,
+    ) -> Result<(), StorageError> {
+        diesel::insert_into(tags::table)
+            .values(
+                snippet
+                    .tags
+                    .iter()
+                    .map(|t| (tags::snippet_id.eq(snippet_id), tags::value.eq(t)))
+                    .collect::<Vec<_>>(),
+            )
+            .on_conflict_do_nothing()
+            .execute(conn)?;
+
+        Ok(())
+    }
+
+    fn trim_removed_tags(
+        &self,
+        conn: &PgConnection,
+        snippet_id: i32,
+        snippet: &Snippet,
+    ) -> Result<(), StorageError> {
+        diesel::delete(
+            tags::table.filter(
+                tags::snippet_id
+                    .eq(snippet_id)
+                    .and(tags::value.ne_all(&snippet.tags)),
+            ),
+        )
+        .execute(conn)?;
+
+        Ok(())
+    }
+}
+
+impl Storage for SqlStorage {
+    fn create(&self, snippet: &Snippet) -> Result<Snippet, StorageError> {
+        let conn = self.pool.get()?;
+        conn.transaction::<_, StorageError, _>(|| {
+            // insert the new snippet row first to get the generated primary key
+            let snippet_id = self.upsert_snippet(&conn, snippet)?;
```

Comment:
**Q:** If it's *upsert*, does it mean it either updates or inserts the record? If so, shouldn't we forbid updating here? Especially when ID conflict is araised?

Reply:
>  If it's upsert, does it mean it either updates or inserts the record?

Correct. It tries to `INSERT` a new row first, and falls back to `UPDATE` in case of a *conflict* (violation of some form of a unique constraint).

>  If so, shouldn't we forbid updating here? Especially when ID conflict is araised?

That's a very good observation! :) As is, the code assumes that duplicate slugs never happen (after all, the probability of such event is tiny: `1 / len(string.ascii_letters + string.digits) ** 8 = 4.580010609233563e-15`), so you are right that an upsert would not work for detection of conflicts, as we simply `UPDATE` the existing row in-place in case of a conflict.

I'm not sure what we should do here :( What I like about the current approach is that it is concise and allows to reuse the very same code for both new insertions and updates. But, as you correctly pointed out, this rules out our initial idea of generating longer IDs on retries.

Reply:
> I'm not sure what we should do here :(

Neither do I . I wonder, what others do in this case? :thinking: One thing is clear for me, we definitely need to drop a comment here explaining a possible gotcha. At least it will show others (potential readers and/or contributors), that it's not a bug but an intentional move.

Reply:
That's fair! I'll update the code.

Reply:
Split it into two separate methods, so that we can properly detect a unique key violation.